compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -4656,7 +4656,12 @@ static bool checkAndReplaceRotation(TR::Node *node,TR::Block *block, TR::Simplif
 
    //traceMsg(TR::comp(), "Spot 2.5 correctLeftShiftAmount = %d mulConst = %d shiftConst = %d sizeof(T)*8 = %d\n",correctLeftShiftAmount,mulConst,shiftConst,sizeof(T)*8);
 
-   if ( (1 << correctLeftShiftAmount) != mulConst )      // Only a rotate if the const values add up to the size of the node type.  ie for ints, this better add up to 32
+   // We can only convert to a rotate if the calculated shift value (based on
+   // operand size) matches the constant from the multiply, specifically the
+   // constant must be 2^correctLeftShiftAmount.
+   // NB the L after the 1 is required for correctness, otherwise 1<<n == 1<<(n+32)
+   // and we fail to distinguish between 32 and 64 bit patterns.
+   if ( (1L << correctLeftShiftAmount) != mulConst )
```

Comment:
Oh. Does that work for both big and little endian?

Reply:
Yes, it does!

The bitshift itself doesn't care about endianness and the source and target memory _must_ have the same endianness because they are both locals.

For completeness, I'll add that the only time when endianness is a concern with `memcpy` is when it's used to copy memory that may have come from a different machine, e.g. copying a byte array read from a network socket or a file on the disk. In those cases, you do have to care about endianness because the byte arrays may have been written by systems with a different endianness than the system that's reading the arrays.

Reply:
Is the reason for the memcpy that unsigned long long and T may have different sizes? Yes, that must be true as we can be called for T as int. Using memcpy as above doesn't work on big endian machines for that case - you get the high order bits instead of the needed low order bits.

Reply:
Hm...

Actually, the memcpy does two things: it correctly converts the unsigned value if T is signed and, as you correctly suggested, it's supposed do a proper narrowing conversion if T is smaller than unsigned long long.

But you're right! It won't work on a big endian machine. Urgh.

Reply:
Well, after much thinking and reading, I think the correct fix is:

```c++
    T expectedMulConst = correctLeftShiftAmount < static_cast<T>(sizeof(T) * 8) ? 1ULL << correctLeftShiftAmount : 0;
    if (expectedMulConst != mulConst)
        return false;
```

The `static_cast<T>(sizeof(T) * 8)` part  could safely be changed to something like `static_cast<T>(sizeof(unsigned long long))` as well.

Reply:
Awesome - many thanks!